upgrade.ts

```diff
@@ -0,0 +1,127 @@
+import { PrecompiledDirectiveFactory } from './src/core/util/bundler';
+import { createBootstrapFn } from './src/platform/browser_utils';
+import { ProviderLiteral } from './src/core/di/provider_util';
+import { reflector } from './src/core/reflection/reflection';
+
+/**
+ * `UgradeAdapterRef` controls a hybrid AngularJS v1 / Angular v2 application,
+ * but we don't have a use for it right now so no point in creating an interface for it...
+ */
+export type UgradeAdapterRef = void;
+
+export interface UpgradeAdapter {
+  new (): upgradeAdapterInstance;
+}
+
+export interface upgradeAdapterInstance {
+  /**
+   * Allows Angular v2 Component to be used from AngularJS v1.
+   */
+  downgradeNg2Component(type: Type): Function;
+  /**
+   * Bootstrap a hybrid AngularJS v1 / Angular v2 application.
+   */
+  bootstrap(element: Element, modules?: any[], config?: angular.IAngularBootstrapConfig): UgradeAdapterRef;
+  /**
+   * Adds a provider to the top level environment of a hybrid AngularJS v1 / Angular v2 application.
+   */
+  addProvider(provider: Type | any[] | any): void;
+  /**
+   * Allows Angular v2 service to be accessible from AngularJS v1.
+   */
+  downgradeNg2Provider(token: any): Function;
+}
+
+export class NgMetadataUpgradeAdapter {
+
+  bootstrap: Function;
+
+  private _upgradeAdapter: upgradeAdapterInstance;
+
+  constructor( UpgradeAdapter: UpgradeAdapter ) {
+    /**
+     * Manage the @angular/upgrade singleton
+     */
+    this._upgradeAdapter = new UpgradeAdapter();
+    /**
+     * Used to bootstrap a hybrid Angular 1 and Angular 2 application,
+     * using the same signature as `bootstrap` from ng-metadata/platform-browser-dynamic
+     *
+     * E.g. `upgradeAdapter.bootstrap(AppComponent, providers)`
+     */
+    this.bootstrap = createBootstrapFn(this._upgradeAdapter.bootstrap.bind(this._upgradeAdapter));
+  }
+
+  /**
+   * Used to register an Angular 2 component as a directive on an Angular 1 module,
+   * where the directive name is automatically created from the selector.
+   *
+   * E.g. `.directive(...upgradeAdapter.downgradeNg2Component(Ng2Component))
+   */
+  downgradeNg2Component( component: Type ): [ string, Function ] {
+    const annotations = reflector.annotations( component );
+    const cmpAnnotation = annotations[ 0 ];
+    const directiveName = resolveDirectiveNameFromSelector( cmpAnnotation.selector );
+    return [ directiveName, this._upgradeAdapter.downgradeNg2Component( component ) ];
+  }
+
+  /**
+   * Used to register an Angular 2 component by including it in the directives array
+   * of an ng-metadata annotated Angular 1 component.
+   *
+   * E.g.
+   * ```
+   * @Component({
+   *  selector: 'foo',
+   *  directives: [upgradeAdapter.provideNg2Component(Ng2Component)],
+   * })
+   * ```
+   */
+  provideNg2Component( component: Type ): PrecompiledDirectiveFactory {
+    const downgradedComponent = this.downgradeNg2Component( component );
```

Comment:
maybe use  destructuring for better readability?

```typescript
const [directiveName,directiveFactory]: [string, PrecompiledDirectiveFactory] = this.downgradeNg2Component(component);
directiveFactory.__precompiledAs = directiveName;
return directiveFactory;
```

Hmm I'm not sure it's good idea to directly add static props on that function. maybe we can use reflect.metadata for this? WDYT?